lib/collection/src/collection_manager/holders/proxy_segment.rs

```diff
@@ -234,13 +234,14 @@ impl ProxySegment {
         {
             let deleted_points = self.deleted_points.upgradable_read();
             if !deleted_points.is_empty() {
+                let mut deleted_points = RwLockUpgradableReadGuard::upgrade(deleted_points);
                 wrapped_segment.with_upgraded(|wrapped_segment| {
                     for point_id in deleted_points.iter() {
                         wrapped_segment.delete_point(op_num, *point_id)?;
                     }
                     OperationResult::Ok(())
                 })?;
```

Comment:
It doesn't seems like there is anything in the flush, which can prevent `with_upgraded` from completion

Reply:
I see now. The locks deceived me :smile: 

In the scenario above, lock 3 and 4 don't depend on others so they are eventually released, preventing the deadlock.